### PR TITLE
Swap timetable entries with arrows

### PR DESCRIPTION
**Pull request by OmeGak**

### CODE REVIEW

File: indico/htdocs/js/indico/jquery/global.js [hunk 1]
```diff
@@ -126,6 +126,12 @@ $(document).ready(function() {
 
                 hide: function(event) {
                     $(this).qtip('destroy');
+                },
+
+                render: function(event, api) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/global.js]: Does this run only once per qtip or every time it's shown?

File: indico/htdocs/js/indico/modules/timetable/timetable/Base.js [hunk 2]
```diff
@@ -1533,6 +1517,15 @@ type("IntervalManagementTimeTable", ["ManagementTimeTable", "IntervalTimeTableMi
         return dfr.promise();
     },
 
+    _removeMissingEntries: function(entries) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/timetable/timetable/Base.js]: same function as above -- different class or accidental duplicate?

File: indico/htdocs/js/indico/modules/timetable/timetable/Draw.js [hunk 3]
```diff
@@ -787,54 +777,51 @@ type("TimetableBlockDisplayMixin",[],
...
-                 self.managementActions.moveEntryUpDown(self.eventData, true);
-                 if (event.stopPropagation) {
-                     event.stopPropagation();
-                 } else {
-                     event.cancelBubble = true;
-                 }
-                 return false;
-             });
+            if (self.eventData.canSwapUp) {
+                arrowUp.attr('title', $T('Move up'));
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/timetable/timetable/Draw.js]: `$T.gettext()`

File: indico/htdocs/js/indico/modules/timetable/timetable/Draw.js [hunk 4]
```diff
@@ -787,54 +777,51 @@ type("TimetableBlockDisplayMixin",[],
...
+                arrowUp.on('click', function(e) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/timetable/timetable/Draw.js]: `e` is unused

File: indico/modules/events/timetable/controllers/legacy.py [hunk 5]
```diff
@@ -315,18 +317,23 @@ def _process(self):
...
-                entries.append(self._schedule(contribution, start_dt))
-        return jsonify(entries=[serialize_entry_update(x) for x in entries])
+        with track_time_changes(auto_extend='end', user=session.user) as changes:
+            for contribution in query:
+                start_dt = find_next_start_dt(contribution.duration,
+                                              obj=self.session_block or self.event_new,
+                                              day=None if self.session_block else day,
+                                              force=True)
+                entry = self._schedule(contribution, start_dt)
+                if entry.end_dt.date() > day:
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/controllers/legacy.py]: convert to event tz before converting to date?

File: indico/modules/events/timetable/controllers/legacy.py [hunk 6]
```diff
@@ -450,16 +457,31 @@ def _process(self):
...
+        else:
+            return SessionManagementLevel.coordinate
+
+    def _checkParams(self, params):
+        RHManageTimetableEntryBase._checkParams(self, params)
+        if self.entry.is_parallel(in_session=self.session is not None):
+            raise BadRequest
+
+    def _process(self):
+        direction = request.form.get('direction')
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/controllers/legacy.py]: is this really optional? if not, `[]`

File: indico/modules/events/timetable/models/entries.py [hunk 7]
```diff
@@ -232,12 +233,23 @@ def end_dt(cls):
...
     @property
+    def session_siblings(self):
+        from indico.modules.events.sessions.models.blocks import SessionBlock
+        if self.type == TimetableEntryType.SESSION_BLOCK:
+            session_id = self.session_block.session_id
+            return self.siblings.filter(TimetableEntry.session_block.has(SessionBlock.session_id == session_id))
+        elif self.parent:
+            return self.siblings
+        else:
+            return TimetableEntry.query.filter(sql.false())
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/models/entries.py]: I think just `False` works too

File: indico/modules/events/timetable/models/entries.py [hunk 8]
```diff
@@ -303,6 +320,17 @@ def move(self, start_dt):
                 child.start_dt += diff
         self.start_dt = start_dt
 
+    def move_next_to(self, sibling, position='before'):
+        if sibling not in self.siblings:
+            raise ValueError
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/models/entries.py]: `('Not a sibling')`

File: indico/modules/events/timetable/models/entries.py [hunk 9]
```diff
@@ -303,6 +320,17 @@ def move(self, start_dt):
...
+        if position not in ('before', 'after'):
+            raise ValueError
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/models/entries.py]: `('Invalid position')`

File: indico/htdocs/js/indico/modules/timetable/timetable/Base.js (hunk 2 above)
Comment (OmeGak) [inline indico/htdocs/js/indico/modules/timetable/timetable/Base.js]: Different logic.

File: indico/htdocs/js/indico/jquery/global.js (hunk 1 above)
Comment (OmeGak) [inline indico/htdocs/js/indico/jquery/global.js]: It runs only on creation time.

File: indico/modules/events/timetable/controllers/legacy.py (hunk 5 above)
Comment (OmeGak) [inline indico/modules/events/timetable/controllers/legacy.py]: We should really add at some point `start_dt_local`/`end_dt_local` autogenerated with a mixin.

File: indico/modules/events/timetable/util.py [hunk 10]
```diff
@@ -332,3 +334,17 @@ def get_time_changes_notifications(changes, tzinfo, entry=None):
...
+@memoize_request
+def get_top_level_entries(event):
+    return event.timetable_entries.filter_by(parent_id=None).all()
+
+
+@memoize_request
+def get_nested_entries(event):
+    entries = event.timetable_entries.filter(TimetableEntry.parent_id.isnot(None)).all()
+    result = defaultdict(list)
+    for x in entries:
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/util.py]: `for entry in entries:` - I'd rather keep `x` and similar variable names inside list/dict/.. comprehensions

File: indico/modules/events/timetable/operations.py [hunk 11]
```diff
@@ -186,3 +186,43 @@ def update_timetable_entry_object(entry, data):
     elif entry.type == TimetableEntryType.BREAK:
         obj.populate_from_dict(data)
     db.session.flush()
+
+
+def swap_timetable_entry(entry, direction, session=None):
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/operations.py]: `session` is shadowing `flask.session` - i'd call it `sess` instead.

**Pull request merged**